CRC failure RX nonce recovery

pkendall64 opened the pull request:
For each non-sync packet the CRC is calculated on the TX with the nonce modulo the FHSS Hop Interval.
On the RX, the packets are decoded and if the CRC fails with the expected none, an attempt is made by trying all the other nonce values within the hop interval and if one succeeds, the nonce is updated to the new value.

Some new "test" code added which shows the effect of updating the CRC with the nonce and what would happen if a false positive is received.

JyeSmith commented:
@pkendall64 I fixed the nonce+crc for all packet types.  Code looks a little better now.

JyeSmith commented:
Ah, you are right.  I'll revert it 👍

JBKingdon commented on src/src/rx_main.cpp:
```diff
@@ -650,23 +652,43 @@ void ICACHE_RAM_ATTR ProcessRFPacket()
-    if (inCRC != calculatedCRC)
+    // artificially inject the nonce on data packets for the CRC calculation
+    int foundNonceIndex = -1;
+    for (int index=0 ; index<3 ; index++)
```
Magic number 3

JBKingdon commented on src/src/rx_main.cpp:
```diff
@@ -650,23 +652,43 @@ void ICACHE_RAM_ATTR ProcessRFPacket()
     {
-        #ifndef DEBUG_SUPPRESS
+        uint8_t tryNonce = (NonceRX + nonceOffset[index]) % ExpressLRS_currAirRate_Modparams->FHSShopInterval;
+        Radio.RXdataBuffer[0] = type | (tryNonce << 2);
```
This Isn't explicit about how many bits are available. It would be easy to mess up with a large fhss interval. Needs some defensive coding or at least a comment

pkendall64 replied:
Yes, this is a bit of a problem.
In common.h it says it can be a max of 16 because we have 4 bits in the sync packet.
In the sync packet code it actually puts the entire 8bits.
On line 317 of tx_main.cpp it assumes that it can only be 2 bits.
And the actual values for the interval are all hard-coded to be 4 in the air-rate tables in common.cpp.
So, theres a bit of tidy up whichever we decide is the right option.

JBKingdon commented on src/src/rx_main.cpp:
```diff
@@ -650,23 +652,43 @@ void ICACHE_RAM_ATTR ProcessRFPacket()
+        uint16_t crc = ota_crc.calc(Radio.RXdataBuffer, 7, CRCInitializer);
+        if (crc == inCRC)
+        {
+            if (NonceRX / ExpressLRS_currAirRate_Modparams->FHSShopInterval == (NonceRX + nonceOffset[index]) / ExpressLRS_currAirRate_Modparams->FHSShopInterval)
```
This conditional confuses me. I'm probably just not awake yet, but assuming the tested offsets are less than fhssinterval it looks like it's just testing for offset 0 and never doing any adjustment. Probably needs a comment to explain what it's actually doing

JBKingdon replied:
Ah nuts,I read it as modulo not divide. Cool.

JBKingdon commented on src/src/tx_main.cpp:
```diff
@@ -354,9 +354,12 @@ void ICACHE_RAM_ATTR SendRCdataToRF()
   ///// Next, Calculate the CRC and put it into the buffer /////
   uint16_t crc = ota_crc.calc(Radio.TXdataBuffer, 7, CRCInitializer);
-  Radio.TXdataBuffer[0] |= (crc >> 6) & 0b11111100;
+  Radio.TXdataBuffer[0] = (Radio.TXdataBuffer[0] & 0b11) | ((crc >> 6) & 0b11111100);
```
Ok, so it's just style, but everyone should be able to understand the use of the combo assignment operators and they provide less opportunity for mistakes when they are appropriate. Recommended for code quality and readability.

schugabe commented:
The test has no assertions and would not fail if something is not working and the tests use random data. This looks more like a verfication for the concept and not an unit test. I don't think that the tests directory is the right place for this in the current form.

pkendall64 commented:
I have split the experimentation code out of the unit tests and added a couple of extra targets that we can use to build and run if we need to collect the data. We can use this for other experimentation code if we need to add any in future.
Perhaps this last commit should be in another PR?
